src/Compilers/CSharp/Portable/Compilation/CSharpDeterministicKeyBuilder.cs

```diff
@@ -0,0 +1,65 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CSharp
+{
+    internal sealed class CSharpDeterministicKeyBuilder : DeterministicKeyBuilder
+    {
+        internal static readonly CSharpDeterministicKeyBuilder Instance = new();
+
+        private CSharpDeterministicKeyBuilder()
+        {
+        }
+
+        protected override void WriteCompilationOptionsCore(JsonWriter writer, CompilationOptions options)
+        {
+            if (options is not CSharpCompilationOptions csharpOptions)
+            {
+                throw new ArgumentException(null, nameof(options));
+            }
+
+            base.WriteCompilationOptionsCore(writer, options);
+
+            writer.Write("unsafe", csharpOptions.AllowUnsafe);
+            writer.Write("topLevelBinderFlags", csharpOptions.TopLevelBinderFlags);
+            writer.WriteKey("usings");
+            writer.WriteArrayStart();
+            foreach (var name in csharpOptions.Usings)
+            {
+                writer.Write(name);
+            }
+            writer.WriteArrayEnd();
+        }
+
+        protected override void WriteParseOptionsCore(JsonWriter writer, ParseOptions parseOptions)
+        {
+            if (parseOptions is not CSharpParseOptions csharpOptions)
+            {
+                throw new ArgumentException(null, nameof(parseOptions));
+            }
+
+            base.WriteParseOptionsCore(writer, parseOptions);
+
+            writer.Write("languageVersion", csharpOptions.LanguageVersion);
+            writer.Write("specifiedLanguageVersion", csharpOptions.SpecifiedLanguageVersion);
+
+            writer.WriteKey("preprocessorSymbols");
+            writer.WriteArrayStart();
+
+            // Even though tools like the command line parser don't explicitly order the symbols 
+            // here the order doesn't actually impact determinism.
+            foreach (var symbol in csharpOptions.PreprocessorSymbols.OrderBy(StringComparer.Ordinal))
```

Comment:
Would it be better to keep the original order then (easier to compare with command line)?

Reply:
The original order isn't recoverable because the command line uses non-ordered collections. I looked into changing it to use ordered collections but it is a pretty significant change. It doesn't really positively impact any other part of the code hence thought it better to just normalize it here. 

Comparing with the original command line isn't really a scenario. Instead it's generally comparing the deterministic key file from two different command lines.